File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int hash[256];
        fill(hash, hash + 256, -1);
        int len=0,maxi=0;
        int i=0,j=0;
        if(s.length()==0) return 0;
        while(j<s.length()){
            if(hash[s[j]]!=-1){
                if(hash[s[j]]>=i){
                   i=hash[s[j]]+1;
                }
            }
            len=j-i+1;
            maxi=max(maxi,len);
            hash[s[j]]=j;
            j++;
        }
        return maxi;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        freq={}
        l=0
        r=0
        ans=0
        for r in range (len(s)):
            while s[r] in freq:
                del freq[s[l]]
                l+=1
            
            freq[s[r]]=1

            ans=max(ans,r-l+1)
        
        return ans